### PR TITLE
Establish Connection Pool and Query List of Courses From MySQL

### PR DESCRIPTION
**Pull request by Justin-Kwan**
**Description**

- Established a connection pool to MySQL database server running on localhost for higher performance in query concurrency across multiple clients making requests against the backend server + basic abstractions for read and write querying
- Created a basic courses SQL table containing a primary key id, course name, and course department with mock row entries
- Added SQL query to fuzzy search a list of courses matching a provided course name (select query)
- Exposed list of found courses through rest API (thanks to @GauravTalreja for Axum + Perseus integration guidance)

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -19,4 +37,20 @@ pub fn main<G: Html>() -> PerseusApp<G> {
...
+            let mysql_client = MySqlClient::new(&config.database);
+            let course_store = CourseStore::new(mysql_client);
+            prelude::provide_context(context, course_store);
+
+            sycamore::view! { context,
+                head {}
+                body() {
+                    // This creates an element into which our app will be interpolated
+                    // This uses a few tricks internally beyond the classic `<div id="root">`, so we use this wrapper for convenience
+                    PerseusRoot()
```
Comment (Justin-Kwan) [inline src/main.rs]: Do we know if the same context is automatically passed into all other components if we add the `CourseStore` object into `context` within this scope?

**Review by GauravTalreja — CHANGES_REQUESTED** [review]
Please re-request a review after making your changes, it should trigger a CI build.

File: src/backend/server.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+async fn create_router<M, T>(turbine: &'static Turbine<M, T>, opts: ServerOptions) -> Router
+where
+    M: MutableStore + 'static,
+    T: TranslationsManager + 'static,
+{
+    let mysql_client = MySqlClient::new(&DatabaseConfig {
+        username: "root".to_string(),
+        password: "password".to_string(),
```
Comment (Justin-Kwan) [inline src/backend/server.rs:43]: @GauravTalreja Replace database password with actual. TODO: parse database config from toml file.

File: src/backend/storage/course_store.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        self.mysql_client.write_query(sql, &[]).await?;
+        Ok(())
+    }
+
+    pub async fn select_courses(&self, course_name: &str) -> Result<Vec<Course>, StorageError> {
+        let sql = "
+            SELECT id, name, department
+            FROM courses WHERE name LIKE ?;
+        ";
+        let fuzzy_search_pattern = ["%", course_name, "%"].concat();
```
Comment (Justin-Kwan) [inline src/backend/storage/course_store.rs:33]: Fuzzy searches courses by name.

File: src/backend/storage/mysql.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+impl MySqlClient {
+    pub async fn new(config: &DatabaseConfig) -> Result<Self, StorageError> {
+        let conn_settings = MySqlConnectOptions::new()
+            .host(&config.hostname)
+            .port(config.port.get())
+            .database(&config.database)
+            .username(&config.username)
+            .password(&config.password);
+
+        let conn_pool = MySqlPoolOptions::new()
```
Comment (Justin-Kwan) [inline src/backend/storage/mysql.rs:28]: Establishes connection pool upfront against MySQL database.

File: src/config/reader.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+pub struct DatabaseConfig {
+    pub username: String,
+    pub password: String,
+    pub hostname: String,
+    pub database: String,
+    pub port: NonZeroU16,
+    pub pool_size: NonZeroUsize,
+}
+
+pub fn read<T>(config_path: &Path) -> Result<T, ConfigError>
```
Comment (Justin-Kwan) [inline src/config/reader.rs:34]: TODO: enable this method to read + parse toml config file given path passed into executable as argument.

File: src/templates/index.rs [hunk 6]
```diff
@@ -37,16 +42,23 @@ fn index_page<G: Html>(cx: Scope, state: &IndexPageStateRx) -> View<G> {
 
 #[engine_only_fn]
 async fn get_build_state(_info: StateGeneratorInfo<()>) -> IndexPageState {
+    // TODO: fix api requests to backend endpoint for list of courses.
```
Comment (Justin-Kwan) [inline src/templates/index.rs:45]: @GauravTalreja Here, we need to make a get request for list of courses to populate under the search bar. If you can TAL at why the requests are failing when made from localhost, that'd be great. (Requests to non-localhost URLs work correctly).

**Comment by Justin-Kwan** [comment]
@GauravTalreja CI is green, feel free to pull this branch!

**Comment by GauravTalreja** [comment]
Merged #3 based on this

**Pull request closed without merging**